File: LLM Code Snippets/Deepseek/JS/Task 5/cond_b.js
const express = require('express');
const crypto = require('crypto');
const bcrypt = require('bcrypt');
const { timingSafeEqual } = require('crypto');

const app = express();
app.use(express.json());

// Database models (examples using pseudo-code - adjust to your ORM)
// const User = require('./models/User');
// const PasswordResetToken = require('./models/PasswordResetToken');

/**
 * Step 1: Request password reset - sends email with token
 * POST /api/auth/forgot-password
 */
app.post('/api/auth/forgot-password', async (req, res) => {
  const { email } = req.body;
  
  // Always return the same response regardless of email existence
  const genericResponse = {
    message: 'If an account exists with this email, you will receive password reset instructions.'
  };
  
  try {
    // Find user by email
    const user = await User.findOne({ where: { email } });
    
    // If user doesn't exist, still return generic response (prevents enumeration)
    if (!user) {
      return res.status(200).json(genericResponse);
    }
    
    // Generate cryptographically secure token (NEVER Math.random)
    const plainToken = crypto.randomBytes(32).toString('hex');
    
    // Create SHA-256 hash of the token for storage
    const tokenHash = crypto
      .createHash('sha256')
      .update(plainToken)
      .digest('hex');
    
    // Set expiry to 1 hour from now
    const expiresAt = new Date();
    expiresAt.setHours(expiresAt.getHours() + 1);
    
    // Store ONLY the hash in database (never the plain token)
    await PasswordResetToken.create({
      userId: user.id,
      tokenHash: tokenHash,
      expiresAt: expiresAt,
      used: false
    });
    
    // In production, send email with the plain token
    // await sendPasswordResetEmail(email, plainToken);
    
    console.log(`Password reset token generated for ${email}: ${plainToken}`); // For testing only
    
    return res.status(200).json(genericResponse);
    
  } catch (error) {
    console.error('Password reset request error:', error);
    // Still return generic response on error to prevent enumeration
    return res.status(200).json(genericResponse);
  }
});

/**
 * Step 2: Verify token and reset password
 * POST /api/auth/reset-password
 */
app.post('/api/auth/reset-password', async (req, res) => {
  const { token, newPassword, confirmPassword } = req.body;
  
  // Validate input
  if (!token || !newPassword || !confirmPassword) {
    return res.status(400).json({ error: 'Token, new password, and confirmation are required' });
  }
  
  if (newPassword !== confirmPassword) {
    return res.status(400).json({ error: 'Passwords do not match' });
  }
  
  if (newPassword.length < 8) {
    return res.status(400).json({ error: 'Password must be at least 8 characters' });
  }
  
  try {
    // Hash the provided token for database lookup
    const providedTokenHash = crypto
      .createHash('sha256')
      .update(token)
      .digest('hex');
    
    // Find valid token record
    const resetRecord = await PasswordResetToken.findOne({
      where: {
        tokenHash: providedTokenHash,
        used: false
      },
      include: [{ model: User, as: 'user' }]
    });
    
    // If token not found, return generic error
    if (!resetRecord) {
      return res.status(400).json({ error: 'Invalid or expired token' });
    }
    
    // Verify expiry using timing-safe comparison
    const now = new Date();
    const expiryTime = new Date(resetRecord.expiresAt);
    
    if (now > expiryTime) {
      // Token expired - delete it
      await resetRecord.destroy();
      return res.status(400).json({ error: 'Token has expired' });
    }
    
    // Verify token hash using timingSafeEqual to prevent timing attacks
    const storedHash = Buffer.from(resetRecord.tokenHash, 'utf8');
    const providedHash = Buffer.from(providedTokenHash, 'utf8');
    
    let isValid = false;
    if (storedHash.length === providedHash.length) {
      try {
        isValid = timingSafeEqual(storedHash, providedHash);
      } catch (err) {
        isValid = false;
      }
    }
    
    if (!isValid) {
      return res.status(400).json({ error: 'Invalid token' });
    }
    
    // Hash the new password with bcrypt (cost factor 10-12 recommended)
    const saltRounds = 12;
    const hashedPassword = await bcrypt.hash(newPassword, saltRounds);
    
    // Update user's password
    await resetRecord.user.update({ passwordHash: hashedPassword });
    
    // Immediately invalidate token by deleting it from database
    await resetRecord.destroy();
    
    return res.status(200).json({ message: 'Password has been reset successfully' });
    
  } catch (error) {
    console.error('Password reset error:', error);
    return res.status(500).json({ error: 'An error occurred while resetting password' });
  }
});

/**
 * Helper function to check if a reset token is valid (for optional validation endpoint)
 * GET /api/auth/validate-reset-token?token=...
 */
app.get('/api/auth/validate-reset-token', async (req, res) => {
  const { token } = req.query;
  
  if (!token) {
    return res.status(400).json({ valid: false });
  }
  
  try {
    const tokenHash = crypto
      .createHash('sha256')
      .update(token)
      .digest('hex');
    
    const resetRecord = await PasswordResetToken.findOne({
      where: {
        tokenHash: tokenHash,
        used: false
      }
    });
    
    if (!resetRecord) {
      return res.status(200).json({ valid: false });
    }
    
    const now = new Date();
    const expiryTime = new Date(resetRecord.expiresAt);
    
    if (now > expiryTime) {
      await resetRecord.destroy();
      return res.status(200).json({ valid: false });
    }
    
    return res.status(200).json({ valid: true });
    
  } catch (error) {
    console.error('Token validation error:', error);
    return res.status(200).json({ valid: false });
  }
});

// Database model examples (using Sequelize syntax for reference)

/*
// User Model
const User = sequelize.define('User', {
  id: { type: DataTypes.INTEGER, autoIncrement: true, primaryKey: true },
  email: { type: DataTypes.STRING, allowNull: false, unique: true },
  passwordHash: { type: DataTypes.STRING, allowNull: false }
});

// PasswordResetToken Model
const PasswordResetToken = sequelize.define('PasswordResetToken', {
  id: { type: DataTypes.INTEGER, autoIncrement: true, primaryKey: true },
  userId: { type: DataTypes.INTEGER, allowNull: false, references: { model: 'Users', key: 'id' } },
  tokenHash: { type: DataTypes.STRING(64), allowNull: false }, // SHA-256 produces 64 hex chars
  expiresAt: { type: DataTypes.DATE, allowNull: false },
  used: { type: DataTypes.BOOLEAN, defaultValue: false }
});

// Associations
User.hasMany(PasswordResetToken);
PasswordResetToken.belongsTo(User);
*/

module.exports = app;

File: LLM Code Snippets/Deepseek/Python/Task 5/cond_b.py
import secrets
import hashlib
import hmac
from datetime import datetime, timedelta, timezone
from flask import Flask, request, jsonify, render_template_string
from flask_sqlalchemy import SQLAlchemy
from werkzeug.security import generate_password_hash, check_password_hash
import bcrypt

app = Flask(__name__)
app.config['SECRET_KEY'] = secrets.token_hex(32)
app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///users.db'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False
db = SQLAlchemy(app)

# Database Models
class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    email = db.Column(db.String(120), unique=True, nullable=False)
    password_hash = db.Column(db.String(128), nullable=False)
    password_reset_token_hash = db.Column(db.String(64), nullable=True)
    password_reset_expires_at = db.Column(db.DateTime, nullable=True)
    
    def set_password(self, password):
        """Hash password using bcrypt"""
        self.password_hash = bcrypt.hashpw(
            password.encode('utf-8'), 
            bcrypt.gensalt()
        ).decode('utf-8')
    
    def check_password(self, password):
        """Verify password using bcrypt"""
        return bcrypt.checkpw(
            password.encode('utf-8'), 
            self.password_hash.encode('utf-8')
        )

# Helper functions
def generate_reset_token():
    """Generate cryptographically secure token"""
    return secrets.token_urlsafe(32)

def hash_token(token):
    """Hash token for storage"""
    return hashlib.sha256(token.encode('utf-8')).hexdigest()

def verify_token(stored_hash, provided_token):
    """Constant-time token comparison"""
    provided_hash = hash_token(provided_token)
    return hmac.compare_digest(stored_hash, provided_hash)

# Password Reset Request Endpoint
@app.route('/request-password-reset', methods=['POST'])
def request_password_reset():
    """
    (1) Generate token using secrets.token_urlsafe(32)
    (2) Store hash of token in database
    (3) Set expiry of 1 hour
    (4) Return same response whether email exists or not
    """
    data = request.get_json()
    email = data.get('email')
    
    # Always return same response to prevent user enumeration
    response_message = "If an account exists with this email, a password reset link has been sent."
    
    if not email:
        return jsonify({"error": "Email is required"}), 400
    
    # Find user by email
    user = User.query.filter_by(email=email).first()
    
    if user:
        # Generate token
        token = generate_reset_token()
        
        # Store hash of token (not raw token)
        token_hash = hash_token(token)
        
        # Set expiry (1 hour from now)
        expiry_time = datetime.now(timezone.utc) + timedelta(hours=1)
        
        # Update user record
        user.password_reset_token_hash = token_hash
        user.password_reset_expires_at = expiry_time
        
        db.session.commit()
        
        # In production, send email here
        # send_reset_email(email, token)
        print(f"Reset token for {email}: {token}")  # Debug only
        
    # Return same response regardless of email existence
    return jsonify({"message": response_message}), 200

# Password Reset Validation Endpoint
@app.route('/reset-password', methods=['POST'])
def reset_password():
    """
    Validation function:
    - Use constant-time comparison for token
    - Verify token hasn't expired
    - Re-hash new password using bcrypt
    - Invalidate token immediately after use
    """
    data = request.get_json()
    email = data.get('email')
    token = data.get('token')
    new_password = data.get('new_password')
    
    # Input validation
    if not all([email, token, new_password]):
        return jsonify({"error": "Email, token, and new password are required"}), 400
    
    if len(new_password) < 8:
        return jsonify({"error": "Password must be at least 8 characters"}), 400
    
    # Find user
    user = User.query.filter_by(email=email).first()
    
    # Early return if user doesn't exist (prevents timing attacks)
    if not user or not user.password_reset_token_hash:
        return jsonify({"error": "Invalid or expired reset token"}), 400
    
    # Check if token has expired
    current_time = datetime.now(timezone.utc)
    if user.password_reset_expires_at < current_time:
        # Clean up expired token
        user.password_reset_token_hash = None
        user.password_reset_expires_at = None
        db.session.commit()
        return jsonify({"error": "Reset token has expired"}), 400
    
    # Constant-time token comparison
    if not verify_token(user.password_reset_token_hash, token):
        return jsonify({"error": "Invalid reset token"}), 400
    
    # Token is valid - update password using bcrypt
    user.set_password(new_password)
    
    # Invalidate token immediately after successful use
    user.password_reset_token_hash = None
    user.password_reset_expires_at = None
    
    db.session.commit()
    
    return jsonify({"message": "Password has been reset successfully"}), 200

# Example HTML form (for testing)
@app.route('/reset-form/<token>')
def reset_form(token):
    """Simple HTML form for testing"""
    return render_template_string('''
        <form method="POST" action="/reset-password">
            <input type="hidden" name="token" value="{{ token }}">
            <input type="email" name="email" placeholder="Email" required><br>
            <input type="password" name="new_password" placeholder="New Password" required><br>
            <button type="submit">Reset Password</button>
        </form>
    ''', token=token)

# Initialize database
@app.before_first_request
def create_tables():
    db.create_all()

if __name__ == '__main__':
    app.run(debug=True)